lib/galaxy/objectstore/irods.py

```diff
@@ -0,0 +1,710 @@
+            else:
+                start_time = datetime.now()
+                log.debug("Pushing cache file '%s' of size %s bytes to collection '%s'", source_file, os.path.getsize(source_file), rel_path)
+                
+                # Create sub-collection first
+                self.session.collections.create(collection_path, recurse=True)
+
+                file_content = None
+                with open(source_file, 'rb') as content_file:
+                    file_content = content_file.read()
+                
+                # Write to file in subcollection created above
+                data_obj = self.session.data_objects.create(data_object_path, recurse=True, **options)
+                with data_obj.open('w') as data_obj_fp:
+                    data_obj_fp.write(file_content)
+
+                end_time = datetime.now()
+                log.debug("Pushed cache file '%s' to collection '%s' (%s bytes transfered in %s sec)",
+                          source_file, rel_path, os.path.getsize(source_file), end_time - start_time)
+            return True
+        else:
+            log.error("Tried updating key '%s' from source file '%s', but source file does not exist.",
+                      rel_path, source_file)
+        return False
+
+    def file_ready(self, obj, **kwargs):
+        """
+        A helper method that checks if a file corresponding to a dataset is
+        ready and available to be used. Return ``True`` if so, ``False`` otherwise.
+        """
+        rel_path = self._construct_path(obj, **kwargs)
+        # Make sure the size in cache is available in its entirety
+        if self._in_cache(rel_path):
+            if os.path.getsize(self._get_cache_path(rel_path)) == self._get_size_in_irods(rel_path):
+                return True
+            log.debug("Waiting for dataset %s to transfer from OS: %s/%s", rel_path,
+                      os.path.getsize(self._get_cache_path(rel_path)), self._get_size_in_irods(rel_path))
+        return False
+
+    def exists(self, obj, **kwargs):
+        in_cache = in_irods = False
+        rel_path = self._construct_path(obj, **kwargs)
+
+        # Check cache
+        if self._in_cache(rel_path):
+            in_cache = True
+        # Check iRODS 
+        in_irods = self._data_object_exists(rel_path)
+
+        # dir_only does not get synced so shortcut the decision
+        dir_only = kwargs.get('dir_only', False)
+        base_dir = kwargs.get('base_dir', None)
+        if dir_only:
+            if in_cache or in_irods:
+                return True
+            # for JOB_WORK directory
+            elif base_dir:
+                if not os.path.exists(rel_path):
+                    os.makedirs(rel_path)
+                return True
+            else:
+                return False
+
+        # TODO: Sync should probably not be done here. Add this to an async upload stack?
```

Comment:
Right, exists should be fast since it is tried often and many operations depend on it being reasonably fast.
Is it really necessary to do this here, or does this mostly concern files that will be transferred later in `update_from_file`

Reply:
I agree that exists() should be light weight and fast. I'm trying to understand how something can be in cache and not in iRODS. Can remove this (or comment it out with a comment for the time being).

Reply:
Yeah, I think we can revisit that when it turns out to be necessary.

Reply:
Ok. I'll add a comment to that effect.

Reply:
Never mind, there already is a comment there!

Reply:
I think that was a misunderstanding, what I meant is that this we can't push to the object store here. If additional syncing is necessary we can revisit this later if necessary,

Reply:
Looks good now, but can you drop the TODO statement in line 466 as well ?